File: python/open_guardrail/guards/code_execution_detect.py
"""Detects code execution patterns."""

import re
import time
from typing import List, Optional

from open_guardrail.core import GuardResult

_PATTERNS = {
    "python": [
        re.compile(r"\beval\s*\(", re.I),
        re.compile(r"\bexec\s*\(", re.I),
        re.compile(r"\bos\.system\s*\(", re.I),
        re.compile(r"\bsubprocess\.", re.I),
    ],
    "javascript": [
        re.compile(r"\bnew\s+Function\s*\(", re.I),
        re.compile(r"\bchild_process\b", re.I),
        re.compile(r"\beval\s*\(", re.I),
    ],
    "shell": [
        re.compile(r"\|\s*bash\b", re.I),
        re.compile(r"\|\s*sh\b", re.I),
    ],
}


class _CodeExecutionDetect:
    def __init__(
        self,
        *,
        action: str = "block",
        languages: Optional[List[str]] = None,
    ) -> None:
        self.name = "code-execution-detect"
        self.action = action
        self.languages = languages or list(_PATTERNS.keys())

    def check(self, text: str, stage: str = "input") -> GuardResult:
        start = time.perf_counter()
        matched: List[str] = []
        for lang in self.languages:
            for p in _PATTERNS.get(lang, []):
                m = p.search(text)
                if m:
                    matched.append(m.group()[:60])
        triggered = len(matched) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name=self.name,
            passed=not triggered,
            action=self.action if triggered else "allow",
            message="Code execution detected" if triggered else None,
            latency_ms=round(elapsed, 2),
            details={"matched": matched} if triggered else None,
        )


def code_execution_detect(
    *,
    action: str = "block",
    languages: Optional[List[str]] = None,
) -> _CodeExecutionDetect:
    return _CodeExecutionDetect(action=action, languages=languages)

File: packages/guards/src/code-execution-detect.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface CodeExecutionDetectOptions {
  action: 'block' | 'warn';
  languages?: string[];
}

const LANG_PATTERNS: Record<string, RegExp[]> = {
  javascript: [/\beval\s*\(/, /\bnew\s+Function\s*\(/, /\bsetTimeout\s*\(\s*['"`]/, /\bdocument\.\w+\s*=/, /\bwindow\.\w+\s*\(/],
  python: [/\bexec\s*\(/, /\b__import__\s*\(/, /\bos\.system\s*\(/, /\bsubprocess\.\w+\s*\(/, /\beval\s*\(/],
  shell: [/\b(?:bash|sh|zsh)\s+-c\b/, /\bcurl\s+.*\|\s*(?:bash|sh)/, /\bwget\s+.*-O\s*-\s*\|/, /\bchmod\s+\+x\b/],
  sql: [/\bEXEC(?:UTE)?\s+/i, /\bxp_cmdshell\b/i, /\bDECLARE\s+@/i, /\bINTO\s+OUTFILE\b/i],
};

export function codeExecutionDetect(options: CodeExecutionDetectOptions): Guard {
  const langs = options.languages ?? Object.keys(LANG_PATTERNS);

  return {
    name: 'code-execution-detect',
    version: '0.1.0',
    description: 'Detects code execution patterns across languages',
    category: 'security',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const found: { language: string; pattern: string }[] = [];
      for (const lang of langs) {
        for (const p of LANG_PATTERNS[lang] ?? []) {
          const m = text.match(p);
          if (m) { found.push({ language: lang, pattern: m[0].slice(0, 40) }); break; }
        }
      }
      const triggered = found.length > 0;
      return { guardName: 'code-execution-detect', passed: !triggered, action: triggered ? options.action : 'allow', latencyMs: Math.round(performance.now() - start), details: triggered ? { found } : undefined };
    },
  };
}
